### ISSUE TITLE
eval loss become nan after a single batch

### ISSUE THREAD

**Issue by ThibaultCastells** [jzhang38/TinyLlama#32]
Hello,
I am trying to finetune the model with [the script you provided](https://github.com/jzhang38/TinyLlama/blob/main/sft/finetune.py), on four RTX 3090 GPUs.
However, I was getting a CUDA out of memory issue, so I made the following change:
```
model = AutoModelForCausalLM.from_pretrained(
    args.model_name_or_path,
    device_map=device_map,
    trust_remote_code=args.trust_remote_code
)
model = model.half()
```
It now fits on my gpu, but the training loss becomes 0 after a single batch, and the evaluation loss is nan.
I tried to check the predictions of the model after training, but its output contains nan so it does not work.
What I already tried to solve the issue:
- different hyper-parameters (lr and wd)
- different datasets (alpaca-cleaned and osst1)
- different checkpoints (TinyLlama-1.1B-intermediate-step-240k-503b and TinyLlama-1.1B-step-50K-105b)

But I get the same result every time. I am assuming this is due to the use of float16, since it is the main difference between my code and the original code. Do you have an idea of what is happening, and of what I could do about it?
Thank you!

**Comment by jzhang38** [comment]
Thanks for your question. I think you conflate the idea of **mixed-precision training** and **half-precision training**.
1. In mixed-precision training. model weight is updated as 32-bit while the forward/backward is calculated in 16 bit.
2. In half-precision training, everything is 16-bit.

What you tried to do here falls under the second category, but TinyLlama is pre-trained with **bf-16 mixed precision**. I believe this change caused numerical instability (actually I don't see many people doing option 2). A better solution for your case is to use qlora. 

Another thought:

**Comment by ThibaultCastells** [comment]
Thank you very much for your explanation and suggestion! I will try it, and let you know the result!

> thought: Have you tried reducing the batch size? I don't think a smaller batch size will cause OOM in 24G RAM.

This is controlled by `per_device_train_batch_size`, am I correct? If so, yes I tried but without success unfortunately (I also tried to set `gradient_checkpointing` to True, and `group_by_length` to True as the description says it saves memory).

**Comment by ThibaultCastells** [comment]
I looked into your solution and realized that it is based on Lora idea, which I suppose means that we are not training the full model. I am interested in fine-tuning the full model for my project.

However I tried the [code to train from scratch](https://github.com/jzhang38/TinyLlama/blob/main/pretrain/tinyllama.py) before and I didn't have memory issue. Is there a reason why you didn't reuse the same code for fine-tuning?
